src/org/sosy_lab/java_smt/example/NQueens.java

```diff
@@ -0,0 +1,242 @@
+ *   ...Q
+ *   Q...
+ *   ..Q.
+ * </pre>
+ */
+public class NQueens {
+  private final SolverContext context;
+  final BooleanFormulaManager bmgr;
+  final int n;
+  final Boolean[][] unsolvableBoard = null;
+
+  private NQueens(SolverContext pContext, int n) {
+    context = pContext;
+    bmgr = context.getFormulaManager().getBooleanFormulaManager();
+    this.n = n;
+  }
+
+  public static void main(String... args)
+      throws InvalidConfigurationException, SolverException, InterruptedException {
+    Configuration config = Configuration.defaultConfiguration();
+    LogManager logger = BasicLogManager.create(config);
+    ShutdownNotifier notifier = ShutdownNotifier.createDummy();
+    Solvers solver = Solvers.SMTINTERPOL;
+    try (SolverContext context =
+             SolverContextFactory.createSolverContext(config, logger, notifier, solver)) {
+      try (Scanner sc = new Scanner(System.in, Charset.defaultCharset().name())) {
+        // Takes input from the user for number of queens to be placed
+        System.out.println("Enter the number of Queens to be " + "placed on the board:");
+        int n = sc.nextInt();
+        NQueens myQueen = new NQueens(context, n);
+        Boolean[][] solutions = myQueen.solve();
+        if (solutions == null) {
+          System.out.println("No solutions found.");
+        } else {
+          System.out.println("Solution:");
+          for (Boolean[] pSolution : solutions) {
+            for (int col = 0; col < solutions[0].length; col++) {
+              if (pSolution[col]) {
+                System.out.print("Q ");
+              } else {
+                System.out.print("_ ");
+              }
+            }
+            System.out.println();
+          }
+          System.out.println();
+        }
+      }
+    } catch (InvalidConfigurationException | UnsatisfiedLinkError e) {
+      logger.logUserException(Level.INFO, e, "Solver " + solver + " is not available.");
+    } catch (UnsupportedOperationException e) {
+      logger.logUserException(Level.INFO, e, e.getMessage());
+    }
+  }
+
+  /* prepare symbols: one symbol for each of the N*N cells.*/
+  BooleanFormula[][] getSymbols() {
+    final BooleanFormula[][] symbols = new BooleanFormula[n][n];
+    for (int row = 0; row < n; row++) {
+      for (int col = 0; col < n; col++) {
+        symbols[row][col] = bmgr.makeVariable("q_" + row + "_" + col);
+      }
+    }
+    return symbols;
+  }
+
+  // This method generates a list of rules that represent the constraints for
+  // the N-Queens problem
+  List<BooleanFormula> getRules(BooleanFormula[][] symbols) {
+    List<BooleanFormula> rules = new ArrayList<>();
+    int symbolLength = symbols.length;
+    /* Rule 1: At least one queen per row,
+         or we can say make sure that there are N Queens on the board
+    */
+    for (BooleanFormula[] rowSymbols : symbols) {
+      List<BooleanFormula> clause =
+          new ArrayList<>(Arrays.asList(rowSymbols).subList(0, symbolLength));
```

Comment:
1. Why create an ArrayList from an existing List returned by asList?
2. Why use a sublist and not just use the whole list?

Reply:
I thought using it might make the program more efficient for larger values of n.